k-distribution/k-tutorial/1_basic/02_basics/README.md

```diff
@@ -1 +1,214 @@
+types of productions, but for now we only care about **constructors** and 
+**functions**. Each declaration separated by the `|` operator is individually
+a single production, and the `|` symbol simply groups together productions that
+have the same sort. For example, we could equally have written an identical K
+definition like so:
+
+```k
+module LESSON-02-B
+
+  syntax Color ::= Yellow()
+  syntax Color ::= Blue()
+  syntax Fruit ::= Banana()
+  syntax Fruit ::= Blueberry()
+  syntax Color ::= colorOf(Fruit) [function]
+
+  rule colorOf(Banana()) => Yellow()
+  rule colorOf(Blueberry()) => Blue()
+
+endmodule
+```
+
+Or even:
+
+```k
+module LESSON-02-C
+
+  syntax Color ::= Yellow()
+                 | Blue()
+                 | colorOf(Fruit) [function]
+  syntax Fruit ::= Banana()
+                 | Blueberry()
+
+  rule colorOf(Banana()) => Yellow()
+  rule colorOf(Blueberry()) => Blue()
+
+endmodule
+```
+
+Each of these types of productions named above has the same underlying syntax,
+but context and **attributes** are used to distinguish between the different
+types. Tokens, brackets, lists, macros, aliases, and anywhere productions will
+be covered in a later lesson, but this lesson does introduce us to constructors
+and functions. `Yellow()`, `Blue()`, `Banana()`, and `Blueberry()` are
+constructors. You can think of a constructor like a constructor for an
+algebraic data type, if you're familiar with a functional language. The data
+type itself is the sort that appears on the left of the `::=` operator. Sorts
+in K consist of uppercase identifiers.
+
+Constructors can have arguments, but these ones do not. We will cover the
+syntax of productions in detail in the next lesson, but for now, you can write
+a production with no arguments as an uppercase or lowercase identifier followed
+by the `()` operator.
+
+A function is distinguished from a constructor by the presence of the
+**function** attribute. Attributes appear in a comma separated list between
+square brackets after any sentence, including both productions and rules.
+Various attributes with built-in meanings exist in K and will be discussed
+throughout the tutorial, but the user can also give any sentence a user-defined
+attribute using the same syntax by simply using an identifier for the attribute
+which is not reserved by K.
+
+## Rules, Matching, and Variables
+
+Functions in K are given definitions using rules. A rule begins with the `rule`
+keyword and contains at least one **rewrite operator**. The rewrite operator
+is represented by the syntax `=>`. The rewrite operator is one of the built-in
+productions in K, and we will discuss in more detail how it can be used in
+future lessons, but for now, you can think of a rule as consisting of a 
+**left-hand side** and a **right-hand side**, separated by the rewrite
+operator. On the left-hand side is the name of the function and zero or more
+**patterns**. On the right-hand side is another pattern. The meaning of the
```

Comment:
I don't understand the "zero or more patterns". Isn't it just a single pattern on the LHS? Or do you mean the use of `#Or`?

Reply:
I'm referring to the arguments to the function. It's true that the entire term on the left hand side is itself a pattern, but this is going to be a confusing way of thinking about it for someone whose background is functional programming. If you have a better suggestion on how to phrase it, I'm all ears.

Reply:
Ah OK, then I would just emphasize that the zero or more patterns are _arguments_ to the function.

```suggestion
operator. On the left-hand side is the name of the function and zero or more
**patterns** in the argument positions of the functions. On the right-hand side is another pattern. The meaning of the
```